packages/core/echo/echo/src/QueryResult.ts

```diff
@@ -0,0 +1,75 @@
+//
+// Copyright 2025 DXOS.org
+//
+
+import { type CleanupFn } from '@dxos/async';
+import { type SpaceId } from '@dxos/keys';
+
+import type * as Entity from './Entity';
+import type { Query } from './query';
+
+/**
+ * Individual query result entry.
+ */
+export type Entry<T extends Entity.Unknown = Entity.Unknown> = {
+  id: string;
+
+  spaceId: SpaceId;
+
+  /**
+   * May not be present for remote results.
+   */
+  object?: T;
+
+  match?: {
+    // TODO(dmaretskyi): text positional info.
+
+    /**
+     * Higher means better match.
+     */
+    rank: number;
+  };
+
+  /**
+   * Query resolution metadata.
+   */
+  // TODO(dmaretskyi): Rename to meta?
+  resolution?: {
+    // TODO(dmaretskyi): Make this more generic.
+    source: 'remote' | 'local' | 'index';
+
+    /**
+     * Query resolution time in milliseconds.
+     */
+    time: number;
+  };
+};
+
+export type OneShotResult<T extends Entity.Unknown = Entity.Unknown> = {
+  results: Entry<T>[];
+  objects: T[];
+};
+
+export type SubscriptionOptions = {
+  /**
+   * Fire the callback immediately.
+   */
+  fire?: boolean;
+};
+
+// TODO(burdon): Narrow types.
+export interface QueryResult<T extends Entity.Unknown = Entity.Unknown> {
```

Comment:
I think we should promote query to Query.ts (and Order, Filter) so that the code is more normalized since these are so related. It feels to me that Database, QueryResult and the three above are qualitatively different than all of the others and we might think about eventually considering them as a different entrypoint/subsystem. but let's keep them here for now. (e.g., we might lift the query defs and impl from echo-db) into "@dxos/echo/Query"